website/project/mailing_list.py

```diff
@@ -0,0 +1,140 @@
+    node_id = re.search(r'[a-z0-9]*@', target).group(0)[:-1]
+    node = Node.load(node_id)
+
+    sender_email = find_email(message['From'])
+    sender = get_user(email=sender_email)
+
+    user_is_admin = 'admin' in node.get_permissions(sender)\
+        if sender and node else False
+
+    mail_params = {
+        'to_addr': sender_email,
+        'mail': mails.DISCUSSIONS_EMAIL_REJECTED,
+        'target_address': target,
+        'user': sender,
+        'node_type': node.project_or_component if node else '',
+        'node_url': node.absolute_url if node else '',
+        'is_admin': user_is_admin,
+        'mail_log_class': MailingListEventLog
+    }
+
+    if not sender:
+        reason = MailingListEventLog.UNAUTHORIZED
+    elif not node:
+        reason = MailingListEventLog.NOT_FOUND
+    elif node.is_deleted:
+        reason = MailingListEventLog.DELETED
+    elif sender not in node.contributors:
+        reason = MailingListEventLog.FORBIDDEN
+    elif not node.mailing_enabled:
+        reason = MailingListEventLog.DISABLED
+    else:
+        reason = ''
+
+    if reason:
+        mails.send_mail(reason=reason, **mail_params)
+    else:
+        send_messages(node, sender, message)
+
+    # Create a log of this mailing event
+    reason = reason if reason else MailingListEventLog.OK
+    MailingListEventLog.create_from_event(
+        content=message,
+        status=reason,
+        node=node,
+        email=sender_email,
+        user=sender,
+    )
+
+def get_recipients(node, sender=None):
+    if node.mailing_enabled:
+        subscription = NotificationSubscription.load(to_subscription_key(node._id, 'mailing_list_events'))
+        return [u for u in subscription.email_transactional if not u == sender]
+    return []
+
+def get_unsubscribes(node):
+    # Non-subscribed users not guaranteed to be in subscription.none
+    # Safer to calculate it
+    if node.mailing_enabled:
+        recipients = get_recipients(node)
+        return [u for u in node.contributors if u not in recipients]
+    return []
+
+def send_messages(node, sender, message):
+    recipients = get_recipients(node, sender=sender)
+    mail = mails.DISCUSSIONS_EMAIL_ACCEPTED
+    mail._subject = '{} [via OSF: {}]'.format(
+        message['subject'].split(' [via OSF')[0],  # Fixes reply subj, if node.title changes
+        unescape_entities(node.title)
+    )
+    from_addr = '{0} <{1}>'.format(sender.fullname, address(node._id))
+    context = {
+        'body': message['stripped-text'],
+        'cat': node.category,
+        'node_url': '{}{}'.format(settings.DOMAIN.rstrip('/'), node.url),
+        'node_title': unescape_entities(node.title),
+        'url': '{}{}settings/#configureNotificationsAnchor'.format(settings.DOMAIN.rstrip('/'), node.url)
+    }
+
+    for recipient in recipients:
+        mails.send_mail(
+            to_addr=recipient.username,
+            mail=mail,
+            from_addr=from_addr,
+            **context
+        )
```

Comment:
What happens if an email bounces or results in an auto-reply? I believe it will be relayed back to the list. If two users have auto-replies set up this could cause an infinite loop.